### ISSUE TITLE
Bug in `IsomorphismPartialPermSemigroup` or `SmallerDegreePartialPermRepresentation`

### ISSUE THREAD

**Issue by james-d-mitchell** [semigroups/Semigroups#817]
In the current release version (4.0.0) I get the following:

~~~
gap> S := UniformBlockBijectionMonoid(4);
<inverse block bijection monoid of degree 4 with 3 generators>
gap> map := SmallerDegreePartialPermRepresentation(S);
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `SmallerDegreePartialPermRepresentation' on 1 arguments at /Users/jdm/gap/lib/methsel2.g:249 called from
<function "HANDLE_METHOD_NOT_FOUND">( <arguments> )
 called from read-eval loop at *stdin*:450
type 'quit;' to quit to outer loop
brk>
gap> map1 := IsomorphismPartialPermSemigroup(S);
MappingByFunction( <inverse block bijection monoid of size 131, degree 4 with 3 generators>,
<inverse partial perm monoid of size 131, rank 131 with 3 generators>, function( x ) ... end )
gap> map2 := SmallerDegreePartialPermRepresentation(Range(map1));
MappingByFunction( <inverse partial perm monoid of size 131, rank 131 with 3 generators>, <inverse partial perm
 semigroup of rank 10 with 3 generators>, function( x ) ... end, function( x ) ... end )
gap> map := CompositionMapping(map2, map1);
CompositionMapping( MappingByFunction( <inverse partial perm monoid of size 131, rank 131 with 3 generators>,
<inverse partial perm semigroup of rank 10 with 3 generators>, function( x ) ... end, function( x ) ... end ),
 MappingByFunction( <inverse block bijection monoid of size 131, degree 4 with 3 generators>,
<inverse partial perm monoid of size 131, rank 131 with 3 generators>, function( x ) ... end ) )
gap> List(GeneratingPairsOfSemigroupCongruence(C), x -> OnTuples(x,  map));
Error, List Element: <list>[4] must have an assigned value in
  AbsInt( w[1] ) at /Users/jdm/gap/pkg/semigroups/gap/main/orbits.gi:167 called from
EvaluateWord( GeneratorsOfSemigroup( T ), Factorization( S, x )
 ) at /Users/jdm/gap/pkg/semigroups/gap/semigroups/semipperm.gi:333 called from
map!.fun( elm ) at /Users/jdm/gap/lib/mapprep.gi:616 called from
ImagesRepresentative( com!.map2, im ) at /Users/jdm/gap/lib/mapprep.gi:286 called from
ImageElm( map, elm ) at /Users/jdm/gap/lib/mapping.gi:863 called from
OnTuples( x, map ) at *stdin*:455 called from
...  at *stdin*:455
type 'quit;' to quit to outer loop
~~~

Two things:

1) The line `List(GeneratingPairsOfSemigroupCongruence(C), x -> OnTuples(x,  map));` should work but doesn't, this is the bug being reported
2) There's no reason that the first call to `SmallerDegreePartialPermRepresentation(S);` shouldn't work (doing what I tried to manually here).

**Comment by wilfwilson** [comment]
What is `C`? Just an arbitrary congruence on `S`?

Note that you can do `map := map1 * map2;` instead of explicitly calling `CompositionMapping`.

**Comment by james-d-mitchell** [comment]
Good question, I’m not 100% sure, it’s certainly a congruence, but I should update this to say what it is precisely

**Comment by james-d-mitchell** [comment]
I can no longer reproduce this error:

~~~
gap>  S := UniformBlockBijectionMonoid(4);
<inverse block bijection monoid of degree 4 with 3 generators>
gap> map1 := IsomorphismPartialPermSemigroup(S);
MappingByFunction( <inverse block bijection monoid of size 131, degree 4 with 3
 generators>, <inverse partial perm monoid of size 131, rank 131 with 3 generators>
 , function( x ) ... end )
gap> map2 := SmallerDegreePartialPermRepresentation(Range(map1));
MappingByFunction( <inverse partial perm monoid of size 131, rank 131 with 3 generators>
 , <inverse partial perm semigroup of rank 10 with 3 generators>
 , function( x ) ... end, function( x ) ... end )
gap> map := CompositionMapping(map2, map1);
CompositionMapping( MappingByFunction( <inverse partial perm monoid of size 131,
 rank 131 with 3 generators>, <inverse partial perm semigroup of rank 10 with 3
 generators>, function( x ) ... end, function( x ) ... end ), MappingByFunction(
<inverse block bijection monoid of size 131, degree 4 with 3 generators>,
<inverse partial perm monoid of size 131, rank 131 with 3 generators>
 , function( x ) ... end ) )
gap> C := SemigroupCongruence(S, [S.1, S.2]);
<2-sided semigroup congruence over <inverse block bijection monoid of size 131,
 degree 4 with 3 generators> with 1 generating pairs>
gap> List(GeneratingPairsOfSemigroupCongruence(C), x -> OnTuples(x,  map));
[ [ (1)(2,3)(4)(5)(6)(7,9)(8,10), <identity partial perm on [ 1, 4, 5, 6 ]> ] ]
gap> C := SemigroupCongruence(S, [S.1, S.2]);
<2-sided semigroup congruence over <inverse block bijection monoid of size 131,
 degree 4 with 3 generators> with 1 generating pairs>
~~~

**Comment by wilfwilson** [comment]
The code that you've just posted also works fine for me.

But (as I mentioned above) the code originally reported in this issue was incomplete, and in particular, no reproducible. More specifically, `C` was never defined, so I'm at least slightly skeptical that the specific issue that was happening there is gone.

**Comment by james-d-mitchell** [comment]
So, there here's an example that demonstrates this bug:

~~~
S := DualSymmetricInverseMonoid(4);
C := SemigroupCongruence(S,  [ Bipartition( [ [ 1, 2, 3, 4, -1, -2, -3, -4 ] ] ),
      Bipartition( [ [ 1, -1 ], [ 2, -2 ], [ 3, -3 ], [ 4, -4 ] ] ) ] );
map := SmallerDegreePartialPermRepresentation(Source(C));
List(GeneratingPairsOfSemigroupCongruence(C), x -> OnTuples(x,  map));
~~~

yields:

~~~
Error, List Element: <list>[4] must have an assigned value in
  AbsInt( w[1] ) at /Users/jdm/gap/pkg/semigroups/gap/main/orbits.gi:167 called from
EvaluateWord( GeneratorsOfSemigroup( T ), Factorization( S, x )
 ) at /Users/jdm/gap/pkg/semigroups/gap/semigroups/semipperm.gi:343 called from
map!.fun( elm ) at /Users/jdm/gap/lib/mapprep.gi:616 called from
ImagesRepresentative( com!.map2, im ) at /Users/jdm/gap/lib/mapprep.gi:286 called from
ImageElm( map, elm ) at /Users/jdm/gap/lib/mapping.gi:863 called from
OnTuples( x, map ) at *stdin*:5 called from
...  at *stdin*:5
type 'quit;' to quit to outer loop
~~~

**Comment by james-d-mitchell** [comment]
I think I've resolved this in #834

**Comment by james-d-mitchell** [comment]
Resolved in v4.0.2
